File: MyCashFlow.Domains/DataObject/Transaction.cs
﻿using System.ComponentModel.DataAnnotations.Schema;
using System;

namespace MyCashFlow.Domains.DataObject
{
	[Table("Transaction", Schema = "Configuration")]
	public class Transaction
	{
		#region Primary key

		public int TransactionID { get; set; }

		#endregion

		#region Standard fields

		public DateTime Date { get; set; }
		public decimal Amount { get; set; }
		public string Note { get; set; }
		public bool Income { get; set; }

		#endregion

		#region Foreign keys

		public int? ProjectID { get; set; }
		public int CreatorID { get; set; }
		public int TransactionTypeID { get; set; }
		public int? PaymentTypeID { get; set; }


		public virtual Project Project { get; set; }
		public virtual User Creator { get; set; }
		public virtual TransactionType TransactionType { get; set; }
		public virtual PaymentType PaymentType { get; set; }

		#endregion
	}
}

File: MyCashFlow.Web/Scripts/MyCashFlow/Models/Transaction/Transaction.ts
class Transaction {
	transactionID: KnockoutObservable<number>;
	date: KnockoutObservable<Date>;
	amount: KnockoutObservable<number>;
	income: KnockoutObservable<boolean>;
	note: KnockoutObservable<string>;
	projectID: KnockoutObservable<number>;
	projects: KnockoutObservableArray<string>;
	transactionTypeID: KnockoutObservable<number>;
	transactionTypes: KnockoutObservableArray<string>;
	paymentMethodID: KnockoutObservable<number>;
	paymentMethods: KnockoutObservableArray<string>;

	constructor(
		transactionID: number,
		date: Date,
		amount: number,
		income: boolean,
		note: string,
		projectID: number,
		transactionTypeID: number,
		paymentMethodID: number) {
		this.transactionID = ko.observable<number>(transactionID);
		this.date = ko.observable<Date>(date);
		this.amount = ko.observable<number>(amount);
		this.income = ko.observable<boolean>(income);
		this.note = ko.observable<string>(note);
		this.projectID = ko.observable<number>(projectID);
		this.projects = ko.observableArray<string>([]);
		this.transactionTypeID = ko.observable<number>(transactionTypeID);
		this.transactionTypes = ko.observableArray<string>([]);
		this.paymentMethodID = ko.observable<number>(paymentMethodID);
		this.paymentMethods = ko.observableArray<string>([]);
	}
}